### PR TITLE
Fix Bug in IntersectCompressedWithLinJump

### LINKED ISSUE #3553: IntersectCompressedWithLinJump fails with data mismatch
The function https://github.com/dgraph-io/dgraph/blob/67231ec0d9a42409492ca93bd699a3b4ca16fcc8/algo/uidlist.go#L72 
 is expected to perform intersection of `uints` inside `dec`, with `v` and put it in `o`. 

But the following test fails on master
```go
func TestIntersectCompressedWithLinJump(t *testing.T) {
	N := 100
	enc := codec.Encoder{BlockSize: 10}
	for i := 0; i < N; i += 10 {
		enc.Add(uint64(i))
	}
	pack := enc.Done()
	dec := codec.Decoder{Pack: pack}

	v := make([]uint64, 0)
	for i := 0; i < N; i += 10 {
		v = append(v, uint64(i))
	}

	o := make([]uint64, 0)
	IntersectCompressedWithLinJump(&dec, v, &o)

	require.Equal(t, v, o)
}
```

### PR DESCRIPTION
**Pull request by animesh2049**
IntersectCompressedWithLinJump function was failing test(issue #3553)
with data mismatch because LinearSeek wasn't behaving properly.
LinearSeek(n) had an expectation that the block with current blockIdx has
already been searched for, but it wasn't being used that way in
IntersectCompressedWithLinJump function. Also LinearSeek was returning
uids of current block, even if next block's base is equal to n.

Now function LinearSeek(n) returns uids of the last block whose base
is less than n. IntersectCompressedWithLinJump stores intersection of
uids in encoded block with an arrary into another array. Fixes #3553

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/3555)

### CODE REVIEW

**Comment by animesh2049** [comment]
___
*[algo/uidlist.go, line 85 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/3555#-LhBln8O533fM1N8a_ip:-LhEfNWyC6fHZRUV_4kE:b-va7ins) ([raw file](https://github.com/dgraph-io/dgraph/blob/e13606a64e760f88fd5cd03b2233bd6817b2bb51/algo/uidlist.go#L85)):*
<details><summary><i>Previously, martinmr (Martin Martinez Rivera) wrote…</i></summary><blockquote>

I don't understand what "more forward" means? is it a typo for "move forward".

If so change the comment to:

// If v[k] isn't in u, move forward.
</blockquote></details>

Done. Yes it was a typo. Fixed it.

**Comment by animesh2049** [comment]
___
*[codec/codec.go, line 192 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/3555#-LhBnUOl78zM0caqLO-w:-LhEgCEC3WKmMjG8gt6s:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/e13606a64e760f88fd5cd03b2233bd6817b2bb51/codec/codec.go#L192)):*
<details><summary><i>Previously, martinmr (Martin Martinez Rivera) wrote…</i></summary><blockquote>

while you are here, can you add a comment documenting this function?
</blockquote></details>

Done.

**Comment by animesh2049** [comment]
___
*[codec/codec_test.go, line 147 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/3555#-LhBmbZ4DWvYLvpVqxwt:-LhEgckI5naTsRVuBvfA:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/e13606a64e760f88fd5cd03b2233bd6817b2bb51/codec/codec_test.go#L147)):*
<details><summary><i>Previously, martinmr (Martin Martinez Rivera) wrote…</i></summary><blockquote>

I think this comment fits in a single line.
</blockquote></details>

Done.

**Comment by animesh2049** [comment]
___
*[codec/codec_test.go, line 113 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/3555#-LhBmWBt6ZE_VcdkH5_U:-LhEpPsYELFh8_BY35LT:b-8ke5zt) ([raw file](https://github.com/dgraph-io/dgraph/blob/e13606a64e760f88fd5cd03b2233bd6817b2bb51/codec/codec_test.go#L113)):*
> Can you come up with a more descriptive name for the tests in here?
I removed the other function. Most of the tests in that were redundant. Named this func `TestLinearSeek`. It's just test for LinearSeek function. What other description is expected in function name ?

**Comment by animesh2049** [comment]
___
*[codec/codec.go, line 192 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/3555#-LhGe4F-8_Y-pkIWo54g:-LhKNIpz3dbd7urRrQRm:b-fdwgd9) ([raw file](https://github.com/dgraph-io/dgraph/blob/918455477b28a63fc435139e296ed046307c0b56/codec/codec.go#L192)):*
<details><summary><i>Previously, mangalaman93 (Aman Mangal) wrote…</i></summary><blockquote>

last -> previous.
And what happens if seek < base of the first block? It would return the first block, which doesn't seem right. Should add a test for this too.
</blockquote></details>

Last seems more appropriate to me here. 
In case when `seek < d.Pack.Blocks[0].Base` and `d.blockIdx == 0`, It will compare `seek` with `d.Pack.Blocks[1].Base` and return uids of the first block. This behavior seems okay to me.  LinearSeek should return uids of the first block having atleast 1 of the uids >= `seek`.

**Comment by animesh2049** [comment]
___
*[codec/codec_test.go, line 127 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/3555#-LhWVHKx3ajBJadY38sJ:-LhZoJnfEt5WC9VYcKmN:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/dae29180e0fc0ce0464c679d360116daedda0c9d/codec/codec_test.go#L127)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

This can be factored out in an internal function. So, you don't need to repeat this logic.
</blockquote></details>

Done.

**Comment by animesh2049** [comment]
___
*[algo/uidlist_test.go, line 456 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/3555#-LhWUJ3T4dcwfjvBQLi-:-LhZoNjjDEptbAQu9Qic:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/dae29180e0fc0ce0464c679d360116daedda0c9d/algo/uidlist_test.go#L456)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

It's very hard to follow what's going on in this test. What's r, what's v, why is r there, what is blkNums? The variable naming needs work. Also, the logic itself needs both comments and simplification/clarification.
</blockquote></details>

Done.

**Comment by animesh2049** [comment]
___
*[algo/uidlist_test.go, line 418 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/3555#-LhcS6a1Bx-qBcKDCIkN:-LhddZidDxVqkZa7uo5E:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f85d7c4830c6a799f354b9b64f09c36156fed5c1/algo/uidlist_test.go#L418)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Remove extra vertical space.
</blockquote></details>

Done.

**Comment by animesh2049** [comment]
___
*[algo/uidlist_test.go, line 471 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/3555#-LhcSvrB-EJf1j-XrwiF:-Lhddmzr2ZZyuQsIZ0iY:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f85d7c4830c6a799f354b9b64f09c36156fed5c1/algo/uidlist_test.go#L471)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

All these sorts should be happening in fillNums.
</blockquote></details>

Done.

**Comment by animesh2049** [comment]
___
*[algo/uidlist_test.go, line 486 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/3555#-LhcSxseBCm4dHKj6PwW:-LhddpJZ7PskmrB2_b5B:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f85d7c4830c6a799f354b9b64f09c36156fed5c1/algo/uidlist_test.go#L486)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Sort should be happening in fillNums.
</blockquote></details>

Done.

**Comment by animesh2049** [comment]
___
*[codec/codec_test.go, line 128 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/3555#-LhcTtTCCZ1HBo4WKAp5:-Lhde8057keOIKz9RzUE:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f85d7c4830c6a799f354b9b64f09c36156fed5c1/codec/codec_test.go#L128)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Is there no require.Contains?
</blockquote></details>

Done.

**Comment by animesh2049** [comment]
___
*[algo/uidlist_test.go, line 467 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/3555#-LhcTQEr3FTu2OfDnrOl:-LhdfY9kAXVt0QnrJ8TM:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f85d7c4830c6a799f354b9b64f09c36156fed5c1/algo/uidlist_test.go#L467)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

s/arrNums/otherNums
</blockquote></details>

Done.

File: algo/uidlist_test.go [hunk 1]
```diff
@@ -412,3 +413,76 @@ func skipDuplicate(in []uint64, idx int) int {
...
+
+	for idx, val := range commonNums {
+		otherNums[idx] = val
+	}
+
+	for idx, val := range extraNums {
+		otherNums[N1+idx] = val
+	}
+
+	sort.Slice(blockNums, func(i, j int) bool { return blockNums[i] < blockNums[j] })
```
Comment (mangalaman93) [inline algo/uidlist_test.go]: can define a function for it.

File: algo/uidlist_test.go [hunk 2]
```diff
@@ -412,3 +413,76 @@ func skipDuplicate(in []uint64, idx int) int {
 	}
 	return i
 }
+
+func fillNums(commonNums, blockNums, otherNums []uint64, N1, N2 int) {
+	extraNums := make([]uint64, N2)
+	for i := range commonNums {
```
Comment (mangalaman93) [inline algo/uidlist_test.go]: you can directly write to the arrays here, would be more readable.

File: algo/uidlist_test.go [hunk 3]
```diff
@@ -412,3 +413,76 @@ func skipDuplicate(in []uint64, idx int) int {
...
+
+			enc := codec.Encoder{BlockSize: 10}
+			for _, num := range blockNums {
+				enc.Add(num)
+			}
+
+			pack := enc.Done()
+			dec := codec.Decoder{Pack: pack}
+
+			o := make([]uint64, 0)
```
Comment (mangalaman93) [inline algo/uidlist_test.go]: call it `actual` or `expected` whichever it is.

**Comment by animesh2049** [comment]
___
*[algo/uidlist_test.go, line 419 at r5](https://reviewable.io/reviews/dgraph-io/dgraph/3555#-LhBENfWTrANVziOhaCN-r5-419:-Lhdsytl6a_EBYWqkpUQ:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/00221128b7968e71dda030ff8238228398ecca3b/algo/uidlist_test.go#L419)):*
<details><summary><i>Previously, mangalaman93 (Aman Mangal) wrote…</i></summary><blockquote>

you can directly write to the arrays here, would be more readable.
</blockquote></details>

Done.

**Comment by animesh2049** [comment]
___
*[algo/uidlist_test.go, line 427 at r5](https://reviewable.io/reviews/dgraph-io/dgraph/3555#-LhBENfWTrANVziOhaCN-r5-427:-Lhdt1dRE0Da3NECP81K:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/00221128b7968e71dda030ff8238228398ecca3b/algo/uidlist_test.go#L427)):*
<details><summary><i>Previously, mangalaman93 (Aman Mangal) wrote…</i></summary><blockquote>

Just u
</blockquote></details>

Done.

**Comment by animesh2049** [comment]
___
*[algo/uidlist_test.go, line 447 at r5](https://reviewable.io/reviews/dgraph-io/dgraph/3555#-LhBENfWTrANVziOhaCN-r5-447:-Lhdt4GDDoXwFmfJ3ALb:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/00221128b7968e71dda030ff8238228398ecca3b/algo/uidlist_test.go#L447)):*
<details><summary><i>Previously, mangalaman93 (Aman Mangal) wrote…</i></summary><blockquote>

can define a function for it.
</blockquote></details>

Done.

**Comment by animesh2049** [comment]
___
*[algo/uidlist_test.go, line 456 at r5](https://reviewable.io/reviews/dgraph-io/dgraph/3555#-LhBENfWTrANVziOhaCN-r5-456:-Lhdt6WwEJorglvntHWU:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/00221128b7968e71dda030ff8238228398ecca3b/algo/uidlist_test.go#L456)):*
<details><summary><i>Previously, mangalaman93 (Aman Mangal) wrote…</i></summary><blockquote>

move this inside `fillNums`
</blockquote></details>

Done.

**Comment by animesh2049** [comment]
___
*[algo/uidlist_test.go, line 468 at r5](https://reviewable.io/reviews/dgraph-io/dgraph/3555#-LhBENfWTrANVziOhaCN-r5-468:-Lhdt9ENAY_19DdJV7dA:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/00221128b7968e71dda030ff8238228398ecca3b/algo/uidlist_test.go#L468)):*
<details><summary><i>Previously, mangalaman93 (Aman Mangal) wrote…</i></summary><blockquote>

doesn't make sense to create these slices here, you can just return them in the function `fillNums`
</blockquote></details>

Done.

**Comment by animesh2049** [comment]
___
*[algo/uidlist_test.go, line 482 at r5](https://reviewable.io/reviews/dgraph-io/dgraph/3555#-LhBENfWTrANVziOhaCN-r5-482:-LhdtBx82FbuOjm2gFGV:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/00221128b7968e71dda030ff8238228398ecca3b/algo/uidlist_test.go#L482)):*
<details><summary><i>Previously, mangalaman93 (Aman Mangal) wrote…</i></summary><blockquote>

call it `actual` or `expected` whichever it is.
</blockquote></details>

Done.

**Comment by animesh2049** [comment]
___
*[codec/codec_test.go, line 131 at r5](https://reviewable.io/reviews/dgraph-io/dgraph/3555#-LhBENfWTrANVziOhaCP-r5-131:-Lhdtnk5BTzL-ebwNB9k:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/00221128b7968e71dda030ff8238228398ecca3b/codec/codec_test.go#L131)):*
<details><summary><i>Previously, mangalaman93 (Aman Mangal) wrote…</i></summary><blockquote>

write a comment here instead of line 135 specifying that expected value of blockIdx here.
</blockquote></details>

Done.
